### PR TITLE
spread headerCellProps into cellProps to use them

### PR DESCRIPTION
**Pull request by keremciu**
I think headerCellProps would be useful to make headerCell. I tried to pass a variable without touching my column object, it's not possible right now.

I think it would be nice to pass all cellProps to cell functions.

### CODE REVIEW

File: src/BaseTable.js [hunk 1]
```diff
@@ -353,7 +353,7 @@ class BaseTable extends React.PureComponent {
     const TableHeaderCell = this._getComponent('TableHeaderCell');
     const SortIndicator = this._getComponent('SortIndicator');
 
-    const cellProps = { columns, column, columnIndex, headerIndex, container: this };
+    const cellProps = { ...headerCellProps, columns, column, columnIndex, headerIndex, container: this };
```
Comment (nihgwu) [inline src/BaseTable.js]: `headerCellProps` could be a function see https://github.com/Autodesk/react-base-table/blob/master/src/BaseTable.js#L372, so we can't just spread it, also it's designed for the cell's container, not the user's cell itself, see https://github.com/Autodesk/react-base-table/blob/master/src/BaseTable.js#L380, so it shouldn't be passed to the renderer

**Comment by nihgwu** [comment]
Hi @keremciu , thanks for your contribution, it's recommended to pass your variables or state to the column definition, or perhaps your cells won't be updated because of the internal optimization

It would be great if you could share some detail of your scenario then I'll try to figure out if there are other ways to achieve that

**Comment by keremciu** [comment]
hello @nihgwu oh thanks for the feedback, it's so useful, I didn't get the cell's container part. yep basically the scenario is:

I put my own headerCell and cell components into table.
```
<BaseTable
  {...props}
  components={{
    TableCell: Cell,
    TableHeaderCell: HeaderCell
  }}
/>
```
I created a util folder called renderers.

it's like ->
- renderers
-- string
-- checkbox
-- contact
-- cta

I put `format: "checkbox"` inside column to call one of these renderers. I'm using same renderers inside many tables.

I wanted to pass general headerProps, rowProps[tableProps] object inside my renderers then I don't need to pass them into every column. Passing into every column is the solution for now. I'm not sure about calling this generic renderers inside column.cellRenderer prop.

**Comment by nihgwu** [comment]
I'd say we have exactly the same scenario to implement `EditableCell` for the inline editing feature. As I pointed out in this section, you could build a advanced `Table` component with your own `components`, and then you could read the `columns` and modify it before passing to `BaseTable`, here is a snippet in our advanced `Table`:
```js
  normalizeColumns(props) {
    const columns = props.columns || normalizeColumns(props.children);
    const columnGroups = props.columnGroups || [];

    this.applyCellRenderer(columns, columnGroups);
```
and part of `applyCellRenderer`
```js
      if (column.editable) {
        column.cellRenderer = this.renderEditableCell;
      } else if (column.type === 'currency') {
        column.cellRenderer = this.renderCurrencyCell;
      } else if (column.type === 'number') {
        column.cellRenderer = this.renderNumberCell;
      } else if (column.type === 'date') {
        column.dataGetter = this.dateGetter;
      }
```

**Comment by nihgwu** [comment]
I just updated the components demo https://autodesk.github.io/react-base-table/examples/components, does that makes sense to you?

The benefit is that the renderers could be used directly via `cellRenderer`, and in the time you could still use `cellRenderer` to override the default result by the format renderer

**Comment by keremciu** [comment]
@nihgwu oh thank you so much, you're great! it's the same approach which I did. The case was I wanted to pass a general prop to renderers. I need to call "onRowSelect" function inside renderer, right now I pass onRowSelect key inside column object then I'm using it inside renderer then I realised I've some props I'm passing to many columns then I wanted to pass them directly to the table then use them inside renderers.

**Comment by nihgwu** [comment]
Maybe it would be related to #11 ?
Not sure why you need the onRowSelect callback in the cerRenderer, I think selectedRowKeys would be enough, and update the state, that’s what we are doing internally 
But I think you could access anything via the container from the renderer callback and it the ref of the table itself, I’d say that’s a bit hacky and I don’t recommend it
I think we could use Context to achieve your needs, and that’s also my current answer to #11 , maybe I would add something like cellState

**Comment by keremciu** [comment]
@nihgwu yes, that's the best case passing them as context to use them inside. it could be maybe theme related props? it would be really useful. 

I think I need to explain again this onRowSelect stuff.

It's like table based object we can pass to the table as prop. It could `extraRenderProps` then I can put some props there like { theme, onRowSelect, onAllRowSelect... } then everyone can put their global stuff there. I understand if you don't wanna add a prop as `extraRenderProps` cause it could maybe change the context of renderers.

I'm gonna close the pull request but it's a really good talk, I hope we can find a solution for #11 and this case

**Comment by nihgwu** [comment]
I'm on my home when I post the above comment, I'd like to explain a bit
you could create the context to hold those props and consume them in custom renderers, and that's my current thoughts on #11 
```jsx
<TableContext.Provider>
  <BaseTable />
</TableContext.Provider>
```
and use it in renderers
```jsx
renderer = props => (
  <TableContext.Consume>
    {context => <div {...props} />}
  </TableContext.Consume>
)
```

also for the other approach I mentioned above, just put everything you want to the table
```
<BaseTable renderers={} />
```
and use in renderers
```jsx
renderer = ({ container }) => {
  const { renderers } = container.props
  return <div {...props} />
}
```

**Comment by nihgwu** [comment]
I designed the component for flexibility. Looking forward your use case
BTW I posted a screenshot of our internal table component built on it today https://github.com/Autodesk/react-base-table#example, I'll try to shift more features to this repo, either demos or snippets to show it's potential

**Pull request closed without merging**